### ISSUE TITLE
I used the latest n3n_3.4.4-86-gb0b6944-1_amd64.deb and made a docker container and ran the tcp transfer and reported an error.

### ISSUE THREAD

**Issue by chenxudong2020** [n42n/n3n#60]
This is my docker project https://github.com/chenxudong2020/n2n-dcoker
In supernode, I only reported missing tcp ports, as follows:
docker run -itd --name=n2n --restart=always -p 1234:1234/tcp -v /usr/n3n/:/etc/n3n/ registry.cn-hangzhou.aliyuncs.com/dubux/n2n:v3

the edge configuration file is as follows:

[tuntap]
name=n3nedge
address=100.100.100.2
address_mode=static
macaddr=DE:AD:BE:EF:99:99
[filter]
allow_routing=true

[community]
name=xxx
key=yyy
supernode=xyz:1234

[connection]
bind=50001
allow_p2p=false
connect_tcp=true

After starting, it was found that the connection was successful, and then the following error was reported:
![image](https://github.com/user-attachments/assets/e12fdb87-fd91-4c49-9936-be60e914234e)
It can be accessed between edges, but it keeps reporting errors and is unstable.

**Comment by hamishcoleman** [comment]
What version supernode are you communicating with?  If you can run the edge with `-vvvv` then we will get more logs.  Also, images are hard to use for debugging.  Finally, can you try the same config without docker to help confirm where the issue might be?

**Comment by chenxudong2020** [comment]
The supernode and edge versions I built using docker are the same. There is no error log in supernode.

**Comment by hamishcoleman** [comment]
Add the `-vvvv` option to the list of paramaters in your CMD statement and that will turn on more debug logging.

Did you try running this without docker?  This will confirm if the problem can be reproduced in a simpler environment that is easier to test

**Comment by chenxudong2020** [comment]
This is the complete log of edge start-vvvv after some time
[2.txt](https://github.com/user-attachments/files/18093734/2.txt)

**Comment by chenxudong2020** [comment]
> Reference

Run the docker command of supernode docker run -itd --name=n2n --restart=always -p 1234:1234/tcp -v /usr/n3n/:/etc/n3n/ registry.cn-hangzhou.aliyuncs.com/dubux /n2n:v3 I only opened the tcp port. I don’t know if it has something to do with it. If I don’t use docker, wouldn’t supernode and edge be used?

**Comment by hamishcoleman** [comment]
The port opening in your docker command is probably not doing what you think it is doing - this is for allowing traffic from the outside world into the container.

If you dont use docker, you can just extract the n3n-edge and run `./n3n-edge start ./edge.conf -vvvv`

**Comment by chenxudong2020** [comment]
Complete log after some time:
[nohup.out (2).txt](https://github.com/user-attachments/files/18094172/nohup.out.2.txt)

**Comment by hamishcoleman** [comment]
Can you try commenting out your `bind=50001` line.

I think it might be causing your issues.  Also, it is unlikely to be achieving a lot in the docker case, since most outbound docker traffic will have a source NAT applied to it

**Comment by chenxudong2020** [comment]
Have you found out what the problem is? I started it locally and checked for a while, but an error also appeared.
![image](https://github.com/user-attachments/assets/b20da560-3448-414c-8c04-2517b56b9eb1)
At this time, I pinged this edge under the window and found that the delay increased as shown below.
![image](https://github.com/user-attachments/assets/b754fb4d-2575-4ee1-8956-d514f1d32445)
Edge under window has the following error, but the frequency is lower than under Linux!
![1733916238924](https://github.com/user-attachments/assets/7c6528ed-a6fb-4cb3-ae4b-d72fec247a9d)

**Comment by hamishcoleman** [comment]
Did you retry with `bind=50001` commented out?

**Comment by hamishcoleman** [comment]
I do not believe I have found the issue - I'm still struggling to reproduce exactly the issue you are seeing.  It has been working fine in almost all my tests - the only hint I could see is that repeated edge reconnections with TCP mode with binding turned on resulted in attempts to reuse the same TCP 4-tuple (which is not a condition supported by TCP stacks)

Running in a docker container sends traffic through an entire set of firewall rules, address translations and pseudo network devices that is different to running on the host - thus it can easily result in a completely different result.

**Comment by chenxudong2020** [comment]
Can't reproduce this problem? I have this problem both when using docker and when not using docker WARNING: supernode not responding, now trying

**Comment by hamishcoleman** [comment]
Yes, I only get the "supernode not responding" when the supernode is actually not responding.

(Excepting the case when I set TCP mode, Bind mode and cause retries to occur)

You should also perform packet captures and check to confirm if there is actual network issues with the connection to the supernode.

**Comment by chenxudong2020** [comment]
![1733918026975](https://github.com/user-attachments/assets/e6edcb78-b3e4-4166-9195-a7c31c608339)
There is no problem with the server. The ping is very stable. Does this mean there is no problem?

**Comment by hamishcoleman** [comment]
Pinging the server is not a measure of the specific issues that might be happening to the actual TCP connection that is causing n3n to report issues - you need to take packet captures and look at the TCP connection to know for sure if there are any issues.

And, no, a stable ping doesnt mean there is no problem.  Which is why I have been trying to replicate your problem

**Comment by chenxudong2020** [comment]
I have no other ideas, the question is on hold for now, waiting for more feedback

**Comment by hamishcoleman** [comment]
It is possible that packet captures would provide that extra information, so that is always a possibility for you to perform.

**Comment by hamishcoleman** [comment]
I've found some more corner cases with the TCP code.  Can you test the latest builds? (If needed, binaries can be downloaded from https://github.com/n42n/n3n/actions/runs/12351950825)
